### PR TITLE
Add .output_shape attribute in all layers (+tests)

### PR DESCRIPTION
**Pull request by matsuyamax**
Note: this PR does NOT modify the model construction API at all, since that is apparently still under debate. Instead, this PR just introduces the logic needed to add automatic shape inference to Keras.

All layers now have an attribute `output_shape` that returns a shape tuple (or a dictionary mapping output names to shapes in the case of the `Graph` container).

The input shape of connected layers is inferred from the output shape of the previous layer. For a layer that is not connected (like the first layer in a network), you can set input shape manually by setting the layer attribute `_input_shape` to a tuple. This allows to skip `Input` layers entirely.

```python
input_layer = Dense(...)
input_layer._input_shape = (None, 10)
```

`None` dimensions in shape tuples are variable dimensions (such as sample number).

I also introduce minor changes in the handling of `stride` for `MaxPooling` layers.

### CODE REVIEW

File: keras/layers/convolutional.py [hunk 1]
```diff
@@ -13,6 +13,37 @@
...
+    return (output_length + stride - 1) // stride
+
+
+def pool_output_length(input_length, pool_size, ignore_border, stride):
+    if ignore_border:
+        output_length = input_length - pool_size + 1
+        output_length = (output_length + stride - 1) // stride
+    else:
+        if pool_size == input_length:
+            output_length = min(input_length, stride - stride % 2)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: `stride - stride % 2`? I find the computations in this function hard to follow in general. Justificative comments would be helpful.

File: keras/layers/convolutional.py [hunk 2]
```diff
@@ -168,14 +214,15 @@ def get_output(self, train):
...
             conv_out = T.nnet.conv.conv2d(X, self.W,
                                           border_mode=border_mode,
                                           subsample=self.subsample)
             if self.border_mode == 'same':
-                shift_x = (self.nb_row - 1) // 2
-                shift_y = (self.nb_col - 1) // 2
-                conv_out = conv_out[:, :, shift_x:X.shape[2] + shift_x, shift_y:X.shape[3] + shift_y]
+                    shift_x = (self.nb_row - 1) // 2
+                    shift_y = (self.nb_col - 1) // 2
+                    conv_out = conv_out[:, :, shift_x:X.shape[2] + shift_x, shift_y:X.shape[3] + shift_y]
```
Comment (fchollet) [inline keras/layers/convolutional.py]: What's up with the whitespace?

**Comment by fchollet** [comment]
One check is failing. One of yours. You should fix that first.

**Comment by matsuyamax** [comment]
> One check is failing. One of yours. You should fix that first.

All tests are passing for me locally with the latest version of Theano. I think this is a matter of upgrading Theano.

**Comment by matsuyamax** [comment]
Tests are passing now after Theano upgrade. Previous failure is a sign that Theano versions before were likely using a buggy implementation of max pooling, and that it is fixed in the current version.

**Comment by fchollet** [comment]
Ok, looks good to me.
